### PR TITLE
Improved logging related to kernels

### PR DESCRIPTION
**Pull request by DonJayamanne**

### CODE REVIEW

File: src/client/common/process/pythonExecutionFactory.ts [hunk 1]
```diff
@@ -59,7 +59,6 @@ export class PythonExecutionFactory implements IPythonExecutionFactory {
     public async create(options: ExecutionFactoryCreationOptions): Promise<IPythonExecutionService> {
         const pythonPath = options.pythonPath ? options.pythonPath : await this.getPythonPath(options.resource);
         const processService: IProcessService = await this.processServiceFactory.create(options.resource);
-        processService.on('exec', this.logger.logProcess.bind(this.logger));
```
Comment (DonJayamanne) [inline src/client/common/process/pythonExecutionFactory.ts:62]: Would result in duplicate logs. we already do this  in the processServivice factory class when calling `create`
Hence this line was unnecessary

File: src/client/datascience/baseJupyterSession.ts [hunk 2]
```diff
@@ -106,12 +106,12 @@ export abstract class BaseJupyterSession implements IJupyterSession {
         this._isDisposed = true;
         if (this.session) {
             try {
-                traceInfo('Shutdown session - current session');
```
Comment (DonJayamanne) [inline src/client/datascience/baseJupyterSession.ts:109]: Turned to verbose for most of these

File: src/client/datascience/jupyter/kernels/helpers.ts [hunk 3]
```diff
@@ -668,7 +668,7 @@ export function findPreferredKernel(
 
                 // Find a kernel spec that matches the language in the notebook metadata.
                 if (score <= 0 && nbMetadataLanguage && speclanguage === (nbMetadataLanguage || '')) {
-                    traceInfo(
+                    traceVerbose(
```
Comment (DonJayamanne) [inline src/client/datascience/jupyter/kernels/helpers.ts:671]: We haven't had issues with kernel selection (at all), hence turning them into verbose.
Reduces the amount of stuff logged

File: src/client/datascience/jupyter/kernels/kernel.ts [hunk 4]
```diff
@@ -500,14 +502,18 @@ export class Kernel implements IKernel {
...
-                traceInfo(
+                traceInfoIfCI(
                     `Kernel got disposed as a result of notebook.onDisposed ${(
                         this.resourceUri || this.notebookDocument.uri
                     ).toString()}`
                 );
-                // this.kernelExecution.cancel();
                 // Ignore when notebook is disposed as a result of failed restarts.
                 if (!this._ignoreNotebookDisposedErrors) {
+                    traceInfo(
```
Comment (DonJayamanne) [inline src/client/datascience/jupyter/kernels/kernel.ts:512]: This was crucial, was difficult to tell why a kernel was disposed.
@rchiodo @IanMatthewHuff FYI - When closing a notebook, VSC un-selects a controller, hence kernels get disposed here.
I wasn't expecting that at all, I was hoping the notebook closed event would get triggered first, but thats not true.

File: src/client/datascience/jupyter/kernels/types.ts [hunk 5]
```diff
@@ -22,10 +22,9 @@ export type LiveKernelModel = IJupyterKernel &
     Partial<IJupyterKernelSpec> & { model: Session.IModel | undefined; notebook?: { path?: string } };
 
 export enum NotebookCellRunState {
-    Running = 1,
-    Idle = 2,
-    Success = 3,
-    Error = 4
+    Idle = 'Idle',
```
Comment (DonJayamanne) [inline src/client/datascience/jupyter/kernels/types.ts:25]: Turned into strings so we can log these, else we just end up with numbers, 2,3,4

File: src/client/datascience/notebook/vscodeNotebookController.ts [hunk 6]
```diff
@@ -237,6 +237,11 @@ export class VSCodeNotebookController implements Disposable {
             // Possible user selected a controller that's not contributed by us at all.
             const kernel = this.kernelProvider.get(event.notebook);
             if (kernel?.kernelConnectionMetadata.id === this.kernelConnection.id) {
+                traceInfo(
+                    `Disposing kernel ${this.kernelConnection.id} for notebook ${getDisplayPath(
+                        event.notebook.uri
+                    )} due to selection of another kernel or closing of the notebook`
+                );
```
Comment (DonJayamanne) [inline src/client/datascience/notebook/vscodeNotebookController.ts:244]: This was crucial, was difficult to tell why a kernel was disposed.
@rchiodo @IanMatthewHuff FYI - When closing a notebook, VSC un-selects a controller, hence kernels get disposed here.
I wasn't expecting that at all, I was hoping the notebook closed event would get triggered first, but thats not true.

File: src/client/datascience/notebook/vscodeNotebookController.ts [hunk 7]
```diff
@@ -345,7 +350,7 @@ export class VSCodeNotebookController implements Disposable {
         );
         const saveKernelInfo = () => {
             const kernelId = kernelSocket?.options.id;
-            if (!kernelId) {
+            if (!kernelId || this.localOrRemoteKernel === 'local') {
```
Comment (DonJayamanne) [inline src/client/datascience/notebook/vscodeNotebookController.ts:353]: We were updating remote kernel ids when dealing with local kernels

File: src/client/datascience/jupyter/kernels/kernel.ts [hunk 8]
```diff
@@ -199,6 +200,7 @@ export class Kernel implements IKernel {
         const sessionPromise = this.startNotebook().then((nb) => nb.session);
         const promise = this.kernelExecution.executeCell(sessionPromise, cell);
         this.trackNotebookCellPerceivedColdTime(stopWatch, sessionPromise, promise).catch(noop);
+        void promise.then((state) => traceInfo(`Cell ${cell.index} executed with state ${state}`));
```
Comment (DonJayamanne) [inline src/client/datascience/jupyter/kernels/kernel.ts:203]: Logging so we know cells completed (today we log when we start cells but not whether it completed or not)
Was crucial for a user issue.

File: src/client/datascience/raw-kernel/rawJupyterSession.ts [hunk 9]
```diff
@@ -360,8 +362,6 @@ export class RawJupyterSession extends BaseJupyterSession {
                         break
         */
 
-        traceWarning(`Didn't get response for requestKernelInfo`);
```
Comment (DonJayamanne) [inline src/client/datascience/raw-kernel/rawJupyterSession.ts:363]: This was wrong

**Review by rchiodo — APPROVED** [review]

**Pull request merged**